pages/docs/manual/latest/variant.mdx

```diff
@@ -193,44 +193,266 @@ let a1 = Warrior({score: 10.5})
+
+It, therefore, compiles to this JS:
+
+```javascript
+var myArray = ["hello", true, false, 13.37];
+```
+
+In the above example, reaching back into the values is as simple as pattern matching on them.
+
+### Pattern matching on nullable values
+
+```rescript
+// The type definition below is inlined here to examplify, but this definition will live in [Core](https://github.com/rescript-association/rescript-core) and be easily accessible
+module Null = {
+  @unboxed type t<'a> = Present('a) | @as(null) Null
+}
+
+type userAge = {ageNum: Null.t<int>}
+
+type rec user = {
+  name: string,
+  age: Null.t<userAge>,
+  bestFriend: Null.t<user>,
+}
+
+let getBestFriendsAge = user =>
+  switch user.bestFriend {
+  | Present({age: Present({ageNum: Present(ageNum)})}) => Some(ageNum)
+  | _ => None
+  }
+```
+
+> Notice how `@as` allows us to say that an unboxed variant case should map to a specific underlying _primitive_. `Present` has a type variable, so it can hold any type. And since it's an unboxed type, only the payloads `'a` or `null` will be kept at runtime. That's where the magic comes from.
+
+### Decoding and encoding JSON idiomatically
+
+With unboxed variants, we have everything we need to define a JSON type:
+
+```rescript
+@unboxed
+type rec json =
+  | @as(false) False
+  | @as(true) True
+  | @as(null) Null
+  | String(string)
+  | Number(float)
+  | Object(Js.Dict.t<json>)
+  | Array(array<json>)
+
+let myValidJsonValue = Array([String("Hi"), Number(123.)])
+```
+
+Here's an example of how you could write your own JSON decoders easily using the above, leveraging pattern matching:
+
+```rescript
+@unboxed
+type rec json =
+  | @as(false) False
+  | @as(true) True
+  | @as(null) Null
+  | String(string)
+  | Number(float)
+  | Object(Js.Dict.t<json>)
+  | Array(array<json>)
+
+type rec user = {
+  name: string,
+  age: int,
+  bestFriend: option<user>,
+}
+
+let rec decodeUser = json =>
+  switch json {
+  | Object(userDict) =>
+    switch (
+      userDict->Dict.get("name"),
+      userDict->Dict.get("age"),
+      userDict->Dict.get("bestFriend"),
+    ) {
+    | (Some(String(name)), Some(Number(age)), Some(maybeBestFriend)) =>
+      Some({
+        name,
+        age: age->Float.toInt,
+        bestFriend: maybeBestFriend->decodeUser,
+      })
+    | _ => None
+    }
+  | _ => None
+  }
+
+let decodeUsers = json =>
+  switch json {
+  | Array(array) => array->Array.map(decodeUser)->Array.keepSome
+  | _ => []
+  }
+```
+
+Encoding that same structure back into JSON is also easy:
+
+```rescript
+let rec userToJson = user => Object(
+  Dict.fromArray([
+    ("name", String(user.name)),
+    ("age", Number(user.age->Int.toFloat)),
+    (
+      "bestFriend",
+      switch user.bestFriend {
+      | None => Null
+      | Some(friend) => userToJson(friend)
+      },
+    ),
+  ]),
+)
+
+let usersToJson = users => Array(users->Array.map(userToJson))
+```
+
+This can be extrapolated to many more cases.
+
+## Coercion
```

Comment:
We should expand this section with:
- Coercing between variants (and the constraints around that)
- Why you can sometimes coerce from variant to string/int/float, and how to think about that (runtime representation must match)
- The last additions of allowing coercing strings to unboxed variants with catch-all string cases

I'll probably think of more. @fhammerschmidt as above, you mind adding a task to the milestone? Feel free to tag me and I can put something together for the actual content.